<a href="https://colab.research.google.com/github/todebode/Food-sales-predictions/blob/main/Food_sales_predictions_Part_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Project 1 - Part 2

Tijesunimi Odebode

02/06/22

In [250]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [251]:
import pandas as pd
filename = "/content/drive/MyDrive/Coding Dojo Bootcamp/Data/sales_predictions.csv"

In [252]:
sales_predictions = pd.read_csv(filename)
sales_predictions.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0           FDA15         9.30  ...  Supermarket Type1          3735.1380
1           DRC01         5.92  ...  Supermarket Type2           443.4228
2           FDN15        17.50  ...  Supermarket Type1          2097.2700
3           FDX07        19.20  ...      Grocery Store           732.3800
4           NCD19         8.93  ...  Supermarket Type1           994.7052

[5 rows x 12 columns]

In [253]:
#1 How many rows and columns?
sales_predictions.shape

(8523, 12)

In [254]:
#2 What are the datatypes of each variable?
sales_predictions.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

In [255]:
#3 Are there duplicates? If so, drop any duplicates.
sales_predictions.duplicated().sum()


0

- There are no duplicates.

In [256]:
#4 Identify missing values.
sales_predictions.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

- There are 1463 missing values in the 'Item_Weight' column.

- There are 2410 missing values in the 'Outlet_Size' column.

In [257]:
#4 This shows the missing values as 'True' in the dataframe
sales_predictions_missing = sales_predictions.isna()
sales_predictions_missing

Item_Identifier  Item_Weight  ...  Outlet_Type  Item_Outlet_Sales
0               False        False  ...        False              False
1               False        False  ...        False              False
2               False        False  ...        False              False
3               False        False  ...        False              False
4               False        False  ...        False              False
...               ...          ...  ...          ...                ...
8518            False        False  ...        False              False
8519            False        False  ...        False              False
8520            False        False  ...        False              False
8521            False        False  ...        False              False
8522            False        False  ...        False              False

[8523 rows x 12 columns]

### #5 Decide on how to address the missing values and do it! (This requires your judgement, so explain your choice)

The following is how I will address the problem:

a) For **"Item_Weight"** column, I will determine the median weight and then insert the median in the cells that have a missing value.

b) For the **"Outlet_Size"** column, 

i) I will count the number of each size (small, medium, High) that I have in the column. 

ii) I will then calculate the average count of all sizes. 

iii) I will fill the cells with missing values with the size whose count is closest to the average.


In [258]:
#5a For "Item_Weight" column, I will determine the median weight and then insert the median in the cells that have a missing value.
sales_item_weight_f = sales_predictions['Item_Weight'].median()
sales_item_weight_d = sales_predictions.loc[:, "Item_Weight"].fillna(sales_item_weight_f, inplace=True)

In [259]:
#5b  For the "Outlet_Size" column,
# i) I will count the number of each size (small, medium, High) that I have in the column.
sales_predictions.Outlet_Size.value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [260]:
# 5b  For the "Outlet_Size" column,
# ii)  I will then calculate the average count of all sizes
Small_size = sales_predictions.Outlet_Size.value_counts().Small 
Medium_size = sales_predictions.Outlet_Size.value_counts().Medium
High_size = sales_predictions.Outlet_Size.value_counts().High 
avg = (Small_size + Medium_size + High_size)/3
avg



2037.6666666666667

In [261]:
# 5b  For the "Outlet_Size" column,
# iii) I will fill the cells with missing values with the size whose count is closest to the average.
sales_predictions.loc[:, "Outlet_Size"].fillna('Small', inplace=True) 
# Since the average value (2037.67) is closest to the number of small sizes in the "Outlet_Size" column, 
# I have filled in 'Small' in the cells with missing values


- For the two columns with missing values in question 5, I used the **"single value imputation"** method to deal with the missing values.I decided not to use the **'deletion'** method because the number of missing values for each column that had a missing value was too many.

In [262]:
#6 Confirm that there are no missing values after addressing them.
sales_predictions.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

- Since all the columns have '0', this confirms that there are no missing values in the dataframe.

In [263]:
#7  Find and fix any inconsistent categories of data (example: fix cat, Cat, and cats so that they are consistent) 
#Let's see the unique values in the 'Item_Fat_Content' column and sort them alphabetically
Item_Fat_Cont = sales_predictions['Item_Fat_Content'].unique()
Item_Fat_Cont.sort()
Item_Fat_Cont

array(['LF', 'Low Fat', 'Regular', 'low fat', 'reg'], dtype=object)

- It is obvious from the unique values that "Low Fat" is written in different ways - capitalized, lower case and even abbreviated. It is also obvious that 'Regular' is written in two ways: 'Regular' and 'reg.'

In [264]:
# We can make all "Low Fat" lowercase and also have a space between 'low' and 'fat'
sales_predictions['Item_Fat_Content'] = sales_predictions['Item_Fat_Content'].replace(['LF','Low Fat'],'low fat')
Item_Fat_Cont = sales_predictions['Item_Fat_Content'].unique()
Item_Fat_Cont.sort()
Item_Fat_Cont

array(['Regular', 'low fat', 'reg'], dtype=object)

In [265]:
# We can make all 'Regular' lowercase as thus: 'regular'
sales_predictions['Item_Fat_Content'] = sales_predictions['Item_Fat_Content'].replace(['Regular','reg'],'regular')
Item_Fat_Cont = sales_predictions['Item_Fat_Content'].unique()
Item_Fat_Cont.sort()
Item_Fat_Cont

array(['low fat', 'regular'], dtype=object)

In [266]:
#7 Let's see the unique values in the 'Item_Type' column and sort them alphabetically
Item_Ty = sales_predictions['Item_Type'].unique()
Item_Ty.sort()
Item_Ty

array(['Baking Goods', 'Breads', 'Breakfast', 'Canned', 'Dairy',
       'Frozen Foods', 'Fruits and Vegetables', 'Hard Drinks',
       'Health and Hygiene', 'Household', 'Meat', 'Others', 'Seafood',
       'Snack Foods', 'Soft Drinks', 'Starchy Foods'], dtype=object)

- It looks like there are no inconsistent categories of data in the 'Item_Type' column.

In [267]:
#7 Let's see the unique values in the 'Outlet_Size' column and sort them alphabetically
Outlet_Si = sales_predictions['Outlet_Size'].unique()
Outlet_Si.sort()
Outlet_Si

array(['High', 'Medium', 'Small'], dtype=object)

- It looks like there are no inconsistent categories of data in the 'Outlet_Size' column.

In [268]:
#7 Let's see the unique values in the 'Outlet_Location_Type' column and sort them alphabetically
Outlet_Loc = sales_predictions['Outlet_Location_Type'].unique()
Outlet_Loc.sort()
Outlet_Loc

array(['Tier 1', 'Tier 2', 'Tier 3'], dtype=object)

- It looks like there are no inconsistent categories of data in the 'Outlet_Location_Type' column.

In [269]:
#7 Let's see the unique values in the 'Outlet_Type' column and sort them alphabetically
Outlet_Ty = sales_predictions['Outlet_Type'].unique()
Outlet_Ty.sort()
Outlet_Ty

array(['Grocery Store', 'Supermarket Type1', 'Supermarket Type2',
       'Supermarket Type3'], dtype=object)

- It looks like there are no inconsistent categories of data in the 'Outlet_Type' column.

In [270]:
#8 For any numerical columns, obtain the summary statistics of each (min, max, mean)
sales_predictions.describe()

Item_Weight  ...  Item_Outlet_Sales
count   8523.00000  ...        8523.000000
mean      12.81342  ...        2181.288914
std        4.22724  ...        1706.499616
min        4.55500  ...          33.290000
25%        9.31000  ...         834.247400
50%       12.60000  ...        1794.331000
75%       16.00000  ...        3101.296400
max       21.35000  ...       13086.964800

[8 rows x 5 columns]